File: sasmodels/models/linear_aggregate.c
/*
 * linear_aggregate
 *
 * Structure factor for a linear aggregate (Debye sum).
 *
 * Parameters:
 *   Q      - momentum transfer
 *   radius_effective - effective scatterer radius (half center-to-center distance)
 *   volfraction - unused (required for P@S parameter order)
 *   n_aggreg  - effective number of points (may be non-integer)
 *
 * Returns:
 *   linear_aggregate(Q)
 *
 * Notes:
 *   Exits with warning if n_aggreg > 100
 */
static double sq_linN(double Q, int n_aggreg, double sep)
{
    int k;
    double SUM, SN;

    double d = sep;
    
    if (n_aggreg <= 1)
        return 1.0;

/*
    if (n_aggreg > 100) {
        fprintf(stderr,
                "WARNING (linear_aggregate): n_aggreg = %d exceeds maximum allowed value (100).\n",
                n_aggreg);
        exit(EXIT_FAILURE);
    }
*/

    SUM = 0.0;

    for (k = 1; k <= n_aggreg - 1; k++) {
        SUM += (double)(n_aggreg - k) * sas_sinx_x(Q * d * (double)k);
    }

    SN = 1.0 + 2.0 * SUM / (double)n_aggreg;

    return SN;
}

/*
 * linear_aggregate (non-integer N)
 *
 * Linear-chain structure factor with non-integer effective length
 * handled by linear interpolation.
 *
 * Parameters:
 *   Q   - momentum transfer
 *   radius_effective - effective scatterer radius (half center-to-center distance)
 *   volfraction - unused
 *   n_aggreg - real (non-integer) number of points
 *
 * Returns:
 *   linear_aggregate(Q)
 */
double Iq(double Q, double radius_effective, double volfraction, double n_aggreg)
{
    int N;
    double w;
    double intensity;

    (void)volfraction;

    if (n_aggreg <= 1.0)
        return 1.0;

    N = (int)n_aggreg;
    w = n_aggreg - (double)N;

    double sep = 2.0 * radius_effective;

    intensity =
        (1.0 - w) * sq_linN(Q, N, sep)
      +        w  * sq_linN(Q, N + 1, sep);

    return intensity;
}

File: sasmodels/models/linear_aggregate.py
r"""
Definition
----------

Linear aggregate structure factor.

WARNING: Should not be used in combination with very anisotropic particle shapes.

Equations are given in
Larsen, A. H., Pedersen, J. S., & Arleth, L. (2020). Assessment of
structure factors for analysis of small-angle scattering data from
desired or undesired aggregates. Applied Crystallography, 53(4), 991-1005.

The model is more specifically eq. (21,22) in the paper.

A non‑integer effective number of points is handled by linear
interpolation between integer chain lengths.

The model evaluates:

.. math::

    S(Q) = (1-w)\,S_N(Q) + w\,S_{N+1}(Q)

where:

- :math:`N = \lfloor RN \rfloor`
- :math:`w = RN - N`

and :math:`S_N(Q)` is the Debye sum for the aggregate.

Parameters
----------
radius_effective : effective scatterer radius (Å), half the center-to-center distance between scatterers on the chain.
volfraction : unused in this :math:`S(q)`; required for :math:`P@S` products.
n_aggreg : aggregation number.

Validation
----------

Translated FORTRAN code

References
----------

# Larsen, A. H., Pedersen, J. S., & Arleth, L. (2020). Assessment of
structure factors for analysis of small-angle scattering data from
desired or undesired aggregates. Applied Crystallography, 53(4), 991-1005.

Authorship and Verification
----------------------------

* **Author:** Jan Skov Pedersen
* **Last Modified by:** Jan Skov Pedersen, April 12, 2026
* **Last Reviewed by:** Reviewer Name Here **Date:**
"""

import numpy as np

name = "linear_aggregate"
title = "Linear aggregate structure factor"
description = """\
Linear aggregate structure factor
"""
category = "structure-factor"
structure_factor = True

parameters = [
    ["radius_effective", "Ang", 10.0, [0.0, np.inf], "",
     "effective scatterer radius (half center-to-center distance)"],
    ["volfraction", "", 0.2, [0.0, 1.0], "",
     "unused in S(q); required for P@S products"],
    ["n_aggreg", "", 50.0, [0.0, 100.0], "", "Aggregation number"],
]

valid = "n_aggreg <= 100"

source = ["linear_aggregate.c"]

def random():
    import random

    return {
        "radius_effective": random.uniform(5, 25),
        "volfraction": random.uniform(0.01, 0.3),
        "n_aggreg": random.uniform(1, 100),
    }

def test():
    print("Structure factor of linear aggregate")
